Python/Product/PythonTools/PythonTools/TokenCache.cs

```diff
@@ -54,13 +53,136 @@ internal struct LineTokenization : ITag {
         }
     }
 
+    internal struct LineToken {
+        public LineToken(TokenInfo token, int lineLength) {
+            Category = token.Category;
+            Trigger = token.Trigger;
+            Column = token.SourceSpan.Start.Column - 1;
+            if (token.SourceSpan.Start.Line == token.SourceSpan.End.Line) {
+                // Token on the same line is easy
+                Length = token.SourceSpan.End.Column - token.SourceSpan.Start.Column;
+            } else if (token.SourceSpan.End.Line == token.SourceSpan.Start.Line + 1 && token.SourceSpan.End.Column == 1) {
+                // Token ending at the start of the next line is a known special case
+                Length = lineLength - Column;
+            } else {
+                // Tokens spanning lines should not be added to a LineTokenization
+                throw new ArgumentException("Cannot cache multiline token");
+            }
+        }
+
+        public TokenCategory Category;
+        public TokenTriggers Trigger;
+        /// <summary>
+        /// 0-based index where the token starts on the line.
+        /// </summary>
+        public int Column;
+        /// <summary>
+        /// Number of characters included in the token.
+        /// </summary>
+        public int Length;
+    }
+
+    public struct TrackingTokenInfo {
```

Comment:
hard to tell from here, does this have to be public?

Reply:
Strictly no, but then we have a stack of internal API on PythonTextBufferInfo. We can do that, but I think there's value in that API being public. But maybe it's better just to hide the whole thing for now...

Reply:
PythonTextBufferInfo isn't public either. I wouldn't worry about making this public if it's not necessary.

Reply:
Ah, you're right. Guess I misread the "sealed" as "public" at some point.